src/shims/unix/wasi/foreign_items.rs

```diff
@@ -0,0 +1,187 @@
+                let len = this.read_machine_usize(len)?;
+
+                this.gen_random(buf.ptr, len)?;
+
+                // Newer versions of the `wasi` crate return a u32, while
+                // older versions return a i16. We support both.
+                let res = match dest.layout.ty {
+                    ty if ty == this.tcx.types.i32 => Scalar::from_i32(0),
+                    ty if ty == this.tcx.types.u16 => Scalar::from_u16(0),
+                    ty => panic!("Unsupported random_get return type: {ty:?}"),
+                };
+                this.write_scalar(res, dest)?;
+            }
+
+            _ => return Ok(EmulateByNameResult::NotSupported),
+        }
+
+        Ok(EmulateByNameResult::NeedsJumping)
+    }
+
+    fn args_get(
+        &mut self,
+        element_heads: &OpTy<'tcx, Provenance>,
+        buffer: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, ()> {
+        let this = self.eval_context_mut();
+
+        let ptr_u8 = this.tcx.mk_mut_ptr(this.tcx.types.u8);
+        let ptr_ptr_u8 = this.tcx.mk_mut_ptr(ptr_u8);
+
+        let ptr_u8_layout = this.layout_of(ptr_u8).unwrap();
+
+        let element_heads = this.addr_to_mplace_ty(element_heads, ptr_ptr_u8)?;
+        let buffer = this.addr_to_mplace_ty(buffer, ptr_u8)?;
+
+        // Based on
+        // https://github.com/bytecodealliance/wasmtime/blob/f85e3f85170bd2c8378977879f4fdf2c6b64de55/crates/wasi-common/src/string_array.rs#L48
+        let mut cursor: u32 = 0;
+        // Can't iterative directly over it to due borrowcheck conflict with 'this'
+        for i in 0..this.machine.args.len() {
+            let arg = &this.machine.args[i];
+            let bytes = arg.as_bytes();
+            let len: u32 = bytes.len().try_into().unwrap();
+            // Make space for `0` terminator.
+            let size = u64::try_from(arg.len()).unwrap().checked_add(1).unwrap();
+
+            let buffer_loc = buffer.ptr.offset(Size::from_bytes(cursor), this)?;
+
+            let arg = OsString::from(arg.to_owned());
+
+            this.write_os_str_to_c_str(&arg, buffer_loc, size)?;
+
+            let head_ptr = element_heads.ptr.offset((i as u64) * ptr_u8_layout.size, this)?;
+            let head_ptr_mplace = MPlaceTy::from_aligned_ptr(head_ptr, ptr_u8_layout);
+
+            this.write_pointer(buffer_loc, &head_ptr_mplace.into())?;
+            cursor = cursor.checked_add(len).unwrap().checked_add(1).unwrap();
+        }
+        Ok(())
+    }
+
+    fn addr_to_mplace_ty(
+        &mut self,
+        arg: &OpTy<'tcx, Provenance>,
+        ptr_ty: Ty<'tcx>,
+    ) -> InterpResult<'tcx, MPlaceTy<'tcx, Provenance>> {
+        let this = self.eval_context_mut();
+        let arg = this.read_scalar(arg)?;
+        //  Hack: New versions of the `wasi` crate perform an int-to-ptr cast before making a syscall,
+        //  while older versions pass a pointer directly. We need to handle both cases.
```

Comment:
Do we have to do this? Can't we make old versions require to use permissive provenance?

Reply:
Wait... you already turn on permissive provenance. Why is this needed *in addition*?

Reply:
We're effectively implementing two different functions here:

* Old:   `pub fn random_get(buf: *mut u8, buf_len: Size) -> Errno;`
* New: `pub fn random_get(arg0: i32, arg1: i32) -> i32;`

If we start out with a pointer, calling `ptr_from_addr_cast(this, arg.to_i32()? as u64)`, won't give us a valid pointer, and we'll later get a `DanglingIntPointer` error.

Reply:
oof... can we just not support the old API instead?

Reply:
wait... the *new* API uses ptr2int and int2ptr roundtrips, and the old one used raw pointers all the way through? What the heck? Where do I need to go complain 😆

Reply:
So... even if I dislike the new API, I think we should only support the new one.

I was unable to find where or why the change happened, do you know?

Reply:
Yeah the new API is clearly not compatible with strict provenance. That's quite terrible. :(

However foreign_items really shouldn't match on `Scalar`. Please add this as a helper method in `intptrcast.rs`.

Please tell me that at least this is a 32bit target, so `to_i32` could also be `to_target_usize`?

Reply:
There is currently no 64bit version of wasi.